app/src/main/java/fr/free/nrw/commons/media/ZoomableActivity.kt

```diff
@@ -0,0 +1,570 @@
+
+    /**
+     * NotForUploadStatus Dao class for database operations
+     */
+    @Inject
+    lateinit var notForUploadStatusDao: NotForUploadStatusDao
+
+    /**
+     * UploadedStatus Dao class for database operations
+     */
+    @Inject
+    lateinit var uploadedStatusDao: UploadedStatusDao
+
+    /**
+     * View Model Factory.
+     */
+    @Inject
+    lateinit var customSelectorViewModelFactory: CustomSelectorViewModelFactory
+
+    /**
+    * Coroutine Dispatchers and Scope.
+    */
+    private var defaultDispatcher : CoroutineDispatcher = Dispatchers.Default
+    private var ioDispatcher : CoroutineDispatcher = Dispatchers.IO
+    private val scope : CoroutineScope = MainScope()
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_zoomable)
+        ButterKnife.bind(this)
+
+        selectedImages = intent.getParcelableArrayListExtra(
+            CustomSelectorConstants.TOTAL_SELECTED_IMAGES
+        )
+        position = intent.getIntExtra(CustomSelectorConstants.PRESENT_POSITION, 0)
+        bucketId = intent.getLongExtra(CustomSelectorConstants.BUCKET_ID, 0L)
+        viewModel = ViewModelProvider(this, customSelectorViewModelFactory).get(
+            CustomSelectorViewModel::class.java
+        )
+        viewModel.fetchImages()
+        viewModel.result.observe(this) {
+            handleResult(it)
+        }
+    }
+
+    /**
+     * Handle view model result.
+     */
+    private fun handleResult(result: Result){
+        if(result.status is CallbackStatus.SUCCESS){
+            val images = result.images
+            if(images.isNotEmpty()) {
+                this@ZoomableActivity.images = ImageHelper.filterImages(images, bucketId)
+                imageUri = if (this@ZoomableActivity.images.isNullOrEmpty()) {
+                    intent.data as Uri
+                } else {
+                    this@ZoomableActivity.images!![position].uri
+                }
+                Timber.d("URL = $imageUri")
+                init(imageUri)
+                onSwipe()
+            }
+        }
+        spinner?.let {
+            it.visibility = if (result.status is CallbackStatus.FETCHING) View.VISIBLE else View.GONE
+        }
+    }
+
+    /**
+     * Handle swap gestures. Ex. onSwipeLeft, onSwipeRight, onSwipeUp, onSwipeDown
+     */
+    private fun onSwipe() {
+        val sharedPreferences: SharedPreferences =
+            getSharedPreferences(ImageHelper.CUSTOM_SELECTOR_PREFERENCE_KEY, 0)
+        val showAlreadyActionedImages =
+            sharedPreferences.getBoolean(ImageHelper.SHOW_ALREADY_ACTIONED_IMAGES_PREFERENCE_KEY, true)
+
+        if (!images.isNullOrEmpty()) {
+            photo!!.setOnTouchListener(object : OnSwipeTouchListener(this) {
+                // Swipe left to view next image in the folder. (if available)
+                override fun onSwipeLeft() {
+                    super.onSwipeLeft()
```

Comment:
Please add `if (photo?.getZoomableController()?.isIdentity() == false) return;` or similar here to not perform the swipe if zoomed in.
Still not perfect, so maybe use `getScaleFactor` and allow for a small margin or something.
Same for the 3 other swipe actions below.